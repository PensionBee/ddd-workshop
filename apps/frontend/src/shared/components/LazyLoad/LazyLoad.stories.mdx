import { Meta, Story, Canvas, ArgsTable } from "@storybook/addon-docs";

<Meta
  title="Shared/Lazy Load/Docs"
  parameters={{
    viewMode: "docs",
    previewTabs: {
      canvas: { hidden: true },
    },
  }}
/>

# Lazy Load

<Canvas>
  <Story id="shared-lazy-load--default" />
</Canvas>
