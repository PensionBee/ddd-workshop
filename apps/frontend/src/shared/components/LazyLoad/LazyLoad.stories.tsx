import type { ComponentStory, ComponentMeta } from "@storybook/react";
import type { LazyLoadProps } from "./LazyLoad";
import { LazyLoad } from "./LazyLoad";

export default {
  title: "Shared/Lazy Load",
  component: LazyLoad,
  parameters: {
    viewMode: "story",
    previewTabs: {
      "storybook/docs/panel": { hidden: true },
    },
  },
} as ComponentMeta<typeof LazyLoad>;

const LazyLoadTemplate: ComponentStory<typeof LazyLoad> = (
  args: LazyLoadProps
) => {
  return <LazyLoad {...args} />;
};

export const Default = LazyLoadTemplate.bind({});
Default.args = {
  children: "Loaded",
  placeholder: "Loading ...",
};
