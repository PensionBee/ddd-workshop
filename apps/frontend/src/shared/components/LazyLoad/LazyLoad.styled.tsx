import styled from "styled-components";

export const StyledLazyLoad = styled.div``;
