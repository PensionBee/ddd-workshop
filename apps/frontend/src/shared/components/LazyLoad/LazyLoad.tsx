import React, { useRef } from "react";
import { useIntersectionObserver } from "@shared/hooks/useIntersectionObserver/useIntersectionObserver";
import { StyledLazyLoad } from "./LazyLoad.styled";

export type LazyLoadProps = {
  /** React children */
  children: React.ReactNode;
  /** Placeholder element */
  placeholder: React.ReactNode;
};

export const LazyLoad: React.FC<LazyLoadProps> = ({
  children,
  placeholder,
  ...props
}) => {
  const lazyContainer = useRef(null);
  const isVisible = useIntersectionObserver(lazyContainer, true);

  if (!isVisible) {
    return (
      <StyledLazyLoad ref={lazyContainer} {...props}>
        {placeholder}
      </StyledLazyLoad>
    );
  }

  return (
    <StyledLazyLoad ref={lazyContainer} {...props}>
      {children}
    </StyledLazyLoad>
  );
};
