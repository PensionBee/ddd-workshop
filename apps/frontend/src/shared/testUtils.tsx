import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
