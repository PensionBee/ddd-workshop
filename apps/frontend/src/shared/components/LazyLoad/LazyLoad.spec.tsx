import { render } from "@shared/testUtils";
import { vi } from "vitest";
import { LazyLoad } from "./LazyLoad";

describe("LazyLoad", () => {
  const lazyContent = "Lazy Content";
  const lazyPlaceholder = "Lazy Placeholder";
  describe("Default", () => {
    it("should render correctly", () => {
      const { getByText, asFragment } = render(
        <LazyLoad placeholder={lazyPlaceholder}>{lazyContent}</LazyLoad>
      );

      expect(getByText(lazyContent)).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
