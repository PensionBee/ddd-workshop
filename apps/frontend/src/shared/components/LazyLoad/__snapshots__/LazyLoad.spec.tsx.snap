// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LazyLoad > Default > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bczSft"
  >
    Lazy Content
  </div>
</DocumentFragment>
`;

exports[`LazyLoad Default should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe"
  >
    Lazy Content
  </div>
</DocumentFragment>
`;
